csharp/src/Ice/TcpConnector.cs

```diff
@@ -72,13 +62,33 @@ public override bool Equals(object? obj)
             }
         }
 
-        public override string ToString() => (_proxy?.Address ?? _addr).ToString()!;
+        public override int GetHashCode()
+        {
+            // This code is thread safe because reading/writing _hashCode (an int) is atomic.
+            if (_hashCode != 0)
+            {
+                // Return cached value
+                return _hashCode;
+            }
+            else
+            {
+                var hash = new System.HashCode();
+                hash.Add(_endpoint.Protocol);
+                hash.Add(_endpoint.Transport);
+                hash.Add(_addr);
+                hash.Add(_endpoint.SourceAddress);
+                _hashCode = hash.ToHashCode();
```

Comment:
You need to check if it's 0 and update it, otherwise a hash code of 0 gets recomputed every time.

Reply:
fixed